### DISCUSSION TITLE
Is it possible to use conditional resolvers in a single form?

### DISCUSSION THREAD

**Discussion by npearson72** [react-hook-form/react-hook-form#9519] [Q&A] [answered]
I would like to have a single form object, but to validate different schemas in a multi-step flow, similar to this:

https://mantine.dev/form/recipes/#form-with-multiple-steps

Can react hook form do this?

If not, what's a good way for validating fields across multiple form objects in a multi-step situation?

**Comment by bluebill1049** [comment]
It's possible to leverage the `context` object.

**Reply by npearson72** [reply]
Hi @bluebill1049, thanks for the reply. Might you be able to suggest how I can use the `FormProvider` / context with multiple forms? I can imagine nesting multiple `FormProvider` components  (one for each form), but how to get the form I need using `useFormContext`?

**Reply by bluebill1049** [reply]
for multiple forms, is it possible to split them into multiple instead of single `useForm` instance?

**Reply by npearson72** [reply]
Yes, that's what I currently have. Multiple `useForm` form objects. I'd like to share/pass their state around, which I guess requires a `FormProvider`, but is it possible to use a single `FormProvider` for multiple `useForm` form objects? If not, and I need to use a 1:1 `FormProvider` per `userForm` object... how can I use the `useFormContext` hook to specify which form in particular I want?

Or maybe I'm just entirely off track?

**Reply by bluebill1049** [reply]
FormProvider is not for tracking multiple form states, you could save your state into a state library such as redux, xstate or zustand. Easier that way.

**Reply by npearson72** [reply]
I see. Is that because react hook form embraces uncontrolled state?

**Reply by bluebill1049** [reply]
It's because each useForm should have its own provider and its own form instance.

```tsx
const formA = useForm;
const formB = useForm;

<StateProvider>
  <FormProvider {...formA}>
    <form />
  </FormProvider>
  
  <FormProvider {...formB}>
    <form />
  </FormProvider>
</StateProvider>
```

it's ok to use without `FormProvider` as well, if you are happy to pass props down.

**Reply by npearson72** [reply]
Is there a way to just create a custom `FormsProvider` and pass multiple form objects to the value prop? Or is the `FormProvider` doing something a bit more magical behind the scenes do deal with the proxy object on `FormState`?

**Comment by Phibee** [comment, marked as answer]
@npearson72, you can check this sample code below I created based on https://mantine.dev/form/recipes/#form-with-multiple-steps. Please let me know if it works. hope it helps.

### Code:
https://codesandbox.io/s/solitary-river-g5ztxx

**Reply by npearson72** [reply]
@Phibee this is gold. Thank you for doing this!

One question: If there are fields in the yup schema that depend on each other when validating (ex: password and passwordConfirmation), will the `useYupValidationResolver` still work?

**Reply by Phibee** [reply]
> @Phibee this is gold. Thank you for doing this!
> 
> One question: If there are fields in the yup schema that depend on each other when validating (ex: password and passwordConfirmation), will the `useYupValidationResolver` still work?

Yes bro, it would still work.
